pkg/sign/signer.go

```diff
@@ -15,29 +15,218 @@
 package sign
 
 import (
+	"bytes"
 	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
 	"crypto/rand"
 	"crypto/sha256"
 	_ "crypto/sha512" // if user chooses SHA2-384 or SHA2-512 for hash
 	"crypto/x509"
 	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
 	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
 
 	protobundle "github.com/sigstore/protobuf-specs/gen/pb-go/bundle/v1"
 	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
 )
 
 type Signer interface {
-	Sign(data []byte) (*protobundle.Bundle, error)
+	Sign(content Content) (*protobundle.Bundle, error)
 }
 
-// TODO: implement
-// type Fulcio struct {
-//	baseURL       string
-//	identityToken string
-// }
+type Fulcio struct {
+	options *FulcioOptions
+}
+
+type FulcioOptions struct {
+	// URL of Fulcio instance
+	BaseURL string
+	// OIDC token to send to Fulcio instance
+	IdentityToken string
+	// Optional timeout for network requests
+	Timeout time.Duration
+	// Optional version string for user agent
+	LibraryVersion string
+}
+
+func SignerFulcio(opts *FulcioOptions) *Fulcio {
+	return &Fulcio{options: opts}
+}
+
+type jsonWebToken struct {
+	Sub string `json:"sub"`
+}
+
+type fulcioCertRequest struct {
+	Credentials      identityToken    `json:"credentials"`
+	PublicKeyRequest publicKeyRequest `json:"publicKeyRequest"`
+}
+
+type identityToken struct {
+	OIDCIdentityToken string `json:"oidcIdentityToken"`
+}
+
+type publicKeyRequest struct {
+	PublicKey         publicKey `json:"publicKey"`
+	ProofOfPossession string    `json:"proofOfPossession"`
+}
+
+type publicKey struct {
+	Algorithm string `json:"algorithm"`
+	Content   string `json:"content"`
+}
+
+type fulcioResponse struct {
+	SCT signedCertificateEmbeddedSct `json:"signedCertificateEmbeddedSct"`
+}
+
+type signedCertificateEmbeddedSct struct {
+	Chain chain `json:"chain"`
+}
+
+type chain struct {
+	Certificates []string `json:"certificates"`
+}
+
+func (f *Fulcio) Sign(content Content) (*protobundle.Bundle, error) {
+	// Get JWT from identity token
+	tokenParts := strings.Split(f.options.IdentityToken, ".")
+	if len(tokenParts) < 2 {
+		return nil, errors.New("Unable to get subject from identity token")
+	}
+
+	jwtString, err := base64.RawURLEncoding.DecodeString(tokenParts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	var jwt jsonWebToken
+	err = json.Unmarshal([]byte(jwtString), &jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate ephemeral keypair
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Can we use the sigstore/sigstore helper `MarshalPublicKeyToPEM`? https://github.com/sigstore/sigstore/blob/main/pkg/cryptoutils/publickey.go#L77. Shouldn't change the dependency tree and just cuts down on some repetition.